commit 9a7ad0fc3b2f8d2bc03e7c2b30a6ddc5532bc14c
Author: haolong.zhang <haolong.zhang@ck-telecom.com>
Date:   Mon Aug 18 12:14:26 2014 +0800

    [Feature] Map key:
        *:Map vol- key(linux keycode) to wifi hostpot key to android space:
        if you  want to use wifi hostspot key just capture "KEYCODE_HOTSPOT" in android space "PhoneWindowManager.java"
        *:Map vol+ key(linux keycode) to camera key to android space:
        if you  want to use camera key just capture "KEYCODE_CAMERA" in android space "PhoneWindowManager.java"
        [id]:N/A
        [TYPE]
           FEATURE
        [Products]
           hicam_baseline
        [Requirement]
           N/A
        [Solution Description]
           N/A
        [Other info]
    	修改:         frameworks/base/api/current.txt
    	修改:         frameworks/base/core/java/android/view/KeyEvent.java
    	修改:         frameworks/base/core/res/res/values/attrs.xml
    	修改:         frameworks/native/include/android/keycodes.h
    	修改:         frameworks/native/include/input/KeycodeLabels.h
    	修改:         frameworks/native/libs/input/Input.cpp
    	修改:         mediatek/config/k95v2_1/mtk-kpd.kl
    
    Change-Id: I510987877ba6086d7f296f85eb18cf97ee971a59

diff --git a/frameworks/base/api/current.txt b/frameworks/base/api/current.txt
index 57db077..7fab2e2 100644
--- a/frameworks/base/api/current.txt
+++ b/frameworks/base/api/current.txt
@@ -26909,6 +26909,7 @@ package android.view {
     field public static final int KEYCODE_VOLUME_DOWN = 25; // 0x19
     field public static final int KEYCODE_VOLUME_MUTE = 164; // 0xa4
     field public static final int KEYCODE_VOLUME_UP = 24; // 0x18
+    field public static final int KEYCODE_HOTSPOT = 231; // 0xe7
     field public static final int KEYCODE_W = 51; // 0x33
     field public static final int KEYCODE_WINDOW = 171; // 0xab
     field public static final int KEYCODE_X = 52; // 0x34
diff --git a/frameworks/base/core/java/android/view/KeyEvent.java b/frameworks/base/core/java/android/view/KeyEvent.java
index 05beeb5..630c928 100755
--- a/frameworks/base/core/java/android/view/KeyEvent.java
+++ b/frameworks/base/core/java/android/view/KeyEvent.java
@@ -666,7 +666,9 @@ public class KeyEvent extends InputEvent implements Parcelable {
     public static final int KEYCODE_GALLERY        = 228;
     
 
-    private static final int LAST_KEYCODE           = KEYCODE_GALLERY;
+	public static final int KEYCODE_HOTSPOT    = 231;
+
+    private static final int LAST_KEYCODE           = KEYCODE_HOTSPOT;
 
     // NOTE: If you add a new keycode here you must also add it to:
     //  isSystem()
@@ -917,6 +919,7 @@ public class KeyEvent extends InputEvent implements Parcelable {
         names.append(KEYCODE_IME_TOGGLE, "KEYCODE_IME_TOGGLE");
         names.append(KEYCODE_MESSAGING, "KEYCODE_MESSAGING");
         names.append(KEYCODE_GALLERY, "KEYCODE_GALLERY");
+		names.append(KEYCODE_HOTSPOT, "KEYCODE_HOTSPOT");
     };
 
     // Symbolic names of all metakeys in bit order from least significant to most significant.
diff --git a/frameworks/base/core/res/res/values/attrs.xml b/frameworks/base/core/res/res/values/attrs.xml
index 42e3b50..109b4ab 100755
--- a/frameworks/base/core/res/res/values/attrs.xml
+++ b/frameworks/base/core/res/res/values/attrs.xml
@@ -1567,6 +1567,7 @@
         <enum name="KEYCODE_BRIGHTNESS_DOWN" value="220" />
         <enum name="KEYCODE_BRIGHTNESS_UP" value="221" />
         <enum name="KEYCODE_MEDIA_AUDIO_TRACK" value="222" />
+	<enum name="KEYCODE_HOTSPOT" value="231" />
     </attr>
 
     <!-- ***************************************************************** -->
diff --git a/frameworks/native/include/android/keycodes.h b/frameworks/native/include/android/keycodes.h
index a8701e1..e108448 100644
--- a/frameworks/native/include/android/keycodes.h
+++ b/frameworks/native/include/android/keycodes.h
@@ -273,6 +273,7 @@ enum {
     AKEYCODE_IME_TOGGLE      = 226,
     AKEYCODE_MESSAGING       = 227,
     AKEYCODE_GALLERY         = 228,
+    AKEYCODE_HOTSPOT    = 231,
     /// @}
 
 
diff --git a/frameworks/native/include/input/KeycodeLabels.h b/frameworks/native/include/input/KeycodeLabels.h
index c1a1268..3caca3b 100644
--- a/frameworks/native/include/input/KeycodeLabels.h
+++ b/frameworks/native/include/input/KeycodeLabels.h
@@ -254,6 +254,7 @@ static const KeycodeLabel KEYCODES[] = {
     { "IME_TOGGLE", 226 },
     { "MESSAGING", 227 },
     { "GALLERY", 228 },
+	{ "HOTSPOT", 231 },
     /// @}
 
     // NOTE: If you add a new keycode here you must also add it to several other files.
diff --git a/frameworks/native/libs/input/Input.cpp b/frameworks/native/libs/input/Input.cpp
index 54fc762..bcddd9b 100644
--- a/frameworks/native/libs/input/Input.cpp
+++ b/frameworks/native/libs/input/Input.cpp
@@ -71,6 +71,7 @@ bool KeyEvent::hasDefaultAction(int32_t keyCode) {
         case AKEYCODE_BRIGHTNESS_DOWN:
         case AKEYCODE_BRIGHTNESS_UP:
         case AKEYCODE_MEDIA_AUDIO_TRACK:
+	case AKEYCODE_HOTSPOT:
             return true;
     }
     
@@ -117,6 +118,7 @@ bool KeyEvent::isSystemKey(int32_t keyCode) {
         case AKEYCODE_IME_TOGGLE:
         case AKEYCODE_MESSAGING:
         case AKEYCODE_GALLERY:
+	case AKEYCODE_HOTSPOT:
         /// @}
             return true;
     }
diff --git a/mediatek/config/ckt95_lw_kk/mtk-kpd.kl b/mediatek/config/ckt95_lw_kk/mtk-kpd.kl
index cdfed53..42ff94c 100755
--- a/mediatek/config/ckt95_lw_kk/mtk-kpd.kl
+++ b/mediatek/config/ckt95_lw_kk/mtk-kpd.kl
@@ -29,11 +29,10 @@ key 103   DPAD_UP
 key 102   HOME
 key 105   DPAD_LEFT
 key 106   DPAD_RIGHT
-key 115   VOLUME_UP         WAKE_DROPPED
-key 114   VOLUME_DOWN       WAKE_DROPPED
+key 115   CAMERA         WAKE_DROPPED
+key 114   HOTSPOT       WAKE_DROPPED
 key 113   MUTE              WAKE_DROPPED
 key 116   POWER             WAKE
-key 212   CAMERA
 key 211   FOCUS
 
 key 16    Q
